### PR TITLE
Add reset solver feature to IC3Base

### PR DESCRIPTION
**Pull request by makaimann**
Adds option to reset solver after certain number of check sat calls (default is 5000).

### CODE REVIEW

**Comment by makaimann** [comment]
@ahmed-irfan here's the reset solver option we discussed before. I haven't tested it on the cluster yet so I marked it do not merge for now. Might be useful for your experiments though. Since it's related to clause subsumption, we can do that in this same PR if you think that makes sense.

File: engines/ic3base.cpp
Comment (ahmed-irfan) [inline engines/ic3base.cpp:719]:
```suggestion
inline Result IC3Base::check_sat()
```

File: engines/ic3base.cpp
Comment (ahmed-irfan) [inline engines/ic3base.cpp:725]:
```suggestion
inline Result IC3Base::check_sat_assuming(const smt::TermVec & assumps)
```

**Review by ahmed-irfan — APPROVED** [review]
LGTM, only two minor comments

**Comment by ahmed-irfan** [comment]
the PR is labeled as `do no merge`. @makaimann Do you plan to add something more in this PR?

**Comment by makaimann** [comment]
Nope, it's ready! Just removed the label. Originally I planned to evaluate it on the cluster before merging but I don't think that's necessary anymore -- especially because it's an option. Instead, we can merge this and https://github.com/upscale-project/pono/pull/146#pullrequestreview-557264063 and evaluate them together.

**Pull request merged**
